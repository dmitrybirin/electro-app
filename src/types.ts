export interface HourConsumptionReal {
  timestamp: number;
  production: number;
  consumption: number;
  losses: null;
  frequency: number;
  system_balance: number;
  ac_balance: number;
  production_renewable: number;
  solar_energy_production: number;
}

export interface HourConsumptionPlanned {
  timestamp: number;
  production: number;
  consumption: number;
  losses: number;
  frequency: number;
  system_balance: number;
  ac_balance: number;
  production_renewable: number;
  solar_energy_forecast: null;
  solar_energy_forecast_operator: null;
}

export interface PlanResult {
  real: HourConsumptionReal[];
  plan: HourConsumptionPlanned[];
}

export interface GraphData {
  hour: number;
  solar: number;
}
