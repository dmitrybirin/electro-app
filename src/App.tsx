import React from 'react';
import { ActivityIndicator, SafeAreaView, StyleSheet, Text } from 'react-native';
import { Chart } from './Chart';
import { usePlanData } from './services/api';

export const App = () => {
  const { planData, loading, errorMessage } = usePlanData();
  return (
    <SafeAreaView style={styles.mainContainer}>
      <Text style={styles.mainText}>Electro</Text>
      <Text style={styles.bodyText}>Planned solar production for the next 24 hours</Text>
      {loading || !planData.plan?.length ? <ActivityIndicator /> : <Chart data={planData} />}
      <Text style={styles.errorMessage}>{errorMessage}</Text>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  mainContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'aqua',
  },
  mainText: {
    fontSize: 24,
  },
  bodyText: {
    fontSize: 16,
  },
  errorMessage: {
    fontSize: 24,
    color: 'red',
  },
  labelTitle: {
    fontSize: 18,
  },
  labelSubTitle: {
    fontSize: 8,
  },
});
