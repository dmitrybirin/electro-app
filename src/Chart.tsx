import React from 'react';
import { StyleSheet } from 'react-native';
import {
  VictoryChart,
  VictoryBar,
  VictoryZoomContainer,
  VictoryAxis,
  VictoryLabel,
} from 'victory-native';

import { getGraphTimeRange, getSolarChartData, formatTimeTicks } from './services/charts';
import { PlanResult } from './types';

interface ChartProps {
  data: PlanResult;
}

export const Chart: React.FC<ChartProps> = ({ data }) => {
  const chartData = getSolarChartData(data);

  return (
    <VictoryChart
      height={500}
      domainPadding={24}
      scale={{ x: 'time', y: 'linear' }}
      containerComponent={
        <VictoryZoomContainer
          allowZoom={false}
          zoomDomain={{ x: getGraphTimeRange() }}
          allowPan={true}
        />
      }>
      <VictoryAxis
        tickLabelComponent={
          <VictoryLabel backgroundPadding={4} style={[styles.labelTitle, styles.labelSubTitle]} />
        }
        tickValues={chartData.map(point => point.timestamp)}
        tickFormat={formatTimeTicks}
      />
      <VictoryAxis dependentAxis />

      <VictoryBar
        barWidth={24}
        data={chartData}
        x="timestamp"
        y="solar"
        animate={{
          onLoad: { duration: 500 },
        }}
      />
    </VictoryChart>
  );
};

const styles = StyleSheet.create({
  labelTitle: {
    fontSize: 18,
  },
  labelSubTitle: {
    fontSize: 8,
  },
});
