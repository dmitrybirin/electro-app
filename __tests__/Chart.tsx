import 'react-native';
import React from 'react';
import { render } from '@testing-library/react-native';
import { Chart } from '../src/Chart';
import { PlanResult } from '../src/types';

const dummyData: PlanResult = {
  real: [
    {
      production: 10,
      consumption: 9,
      losses: null,
      frequency: 42,
      system_balance: 42,
      ac_balance: 42,
      production_renewable: 42,
      solar_energy_production: 500,
      timestamp: +new Date(),
    },
  ],
  plan: [
    {
      production: 10,
      consumption: 9,
      losses: null,
      frequency: 42,
      system_balance: 42,
      ac_balance: 42,
      production_renewable: 42,
      solar_energy_production: 500,
      timestamp: +new Date(),
    },
  ],
};

it('renders chart correctly', () => {
  const { toJSON } = render(<Chart data={dummyData} />);
  expect(toJSON()).toMatchSnapshot();
});
