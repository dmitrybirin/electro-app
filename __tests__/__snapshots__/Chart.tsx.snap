// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders chart correctly 1`] = `
<View
  className="VictoryContainer"
  pointerEvents="box-none"
  style={
    {
      "position": "relative",
    }
  }
>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight="500"
    bbWidth="750"
    focusable={false}
    height={500}
    meetOrSlice={0}
    minX={0}
    minY={0}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchPinch={[Function]}
    onTouchStart={[Function]}
    role="img"
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        {
          "height": "100%",
          "width": "100%",
        },
        {
          "flex": 0,
          "height": 500,
          "width": 750,
        },
      ]
    }
    vbHeight={500}
    vbWidth={750}
    width={750}
  >
    <RNSVGGroup>
      <RNSVGRect
        fill={null}
        height="500"
        propList={
          [
            "fill",
          ]
        }
        width="750"
        x="0"
        y="0"
      />
      <RNSVGGroup>
        <RNSVGLine
          propList={
            [
              "stroke",
            ]
          }
          stroke={
            {
              "payload": 4278190080,
              "type": 0,
            }
          }
          vectorEffect={1}
          x1="0"
          x2="0"
          y1="0"
          y2="0"
        />
      </RNSVGGroup>
      <RNSVGGroup>
        <RNSVGLine
          fill={
            {
              "payload": 0,
              "type": 0,
            }
          }
          propList={
            [
              "fill",
              "stroke",
              "strokeWidth",
              "strokeLinecap",
              "strokeLinejoin",
            ]
          }
          stroke={
            {
              "payload": 4280624421,
              "type": 0,
            }
          }
          strokeLinecap={1}
          strokeLinejoin={1}
          strokeWidth="1"
          vectorEffect={1}
          x1="50"
          x2="700"
          y1="450"
          y2="450"
        />
        <RNSVGGroup>
          <RNSVGText
            content={null}
            dx={
              [
                "0",
              ]
            }
            dy={[]}
            font={{}}
            name="chart-axis-1-tickLabels-0"
            rotate={[]}
            x={
              [
                "175.20630260143935",
              ]
            }
            y={
              [
                "472.97",
              ]
            }
          >
            <RNSVGTSpan
              content="100"
              dx={
                [
                  "0",
                ]
              }
              dy={
                [
                  "0",
                ]
              }
              fill={
                {
                  "payload": 4280624421,
                  "type": 0,
                }
              }
              font={
                {
                  "fontFamily": "Gill Sans",
                  "fontSize": "14",
                  "letterSpacing": "normal",
                  "textAnchor": "middle",
                }
              }
              propList={
                [
                  "fill",
                ]
              }
              rotate={[]}
              x={
                [
                  "175.20630260143935",
                ]
              }
              y={[]}
            />
          </RNSVGText>
        </RNSVGGroup>
        <RNSVGGroup>
          <RNSVGText
            content={null}
            dx={
              [
                "0",
              ]
            }
            dy={[]}
            font={{}}
            name="chart-axis-1-tickLabels-1"
            rotate={[]}
            x={
              [
                "300.41260520287875",
              ]
            }
            y={
              [
                "472.97",
              ]
            }
          >
            <RNSVGTSpan
              content="200"
              dx={
                [
                  "0",
                ]
              }
              dy={
                [
                  "0",
                ]
              }
              fill={
                {
                  "payload": 4280624421,
                  "type": 0,
                }
              }
              font={
                {
                  "fontFamily": "Gill Sans",
                  "fontSize": "14",
                  "letterSpacing": "normal",
                  "textAnchor": "middle",
                }
              }
              propList={
                [
                  "fill",
                ]
              }
              rotate={[]}
              x={
                [
                  "300.41260520287875",
                ]
              }
              y={[]}
            />
          </RNSVGText>
        </RNSVGGroup>
        <RNSVGGroup>
          <RNSVGText
            content={null}
            dx={
              [
                "0",
              ]
            }
            dy={[]}
            font={{}}
            name="chart-axis-1-tickLabels-2"
            rotate={[]}
            x={
              [
                "425.6189078043181",
              ]
            }
            y={
              [
                "472.97",
              ]
            }
          >
            <RNSVGTSpan
              content="300"
              dx={
                [
                  "0",
                ]
              }
              dy={
                [
                  "0",
                ]
              }
              fill={
                {
                  "payload": 4280624421,
                  "type": 0,
                }
              }
              font={
                {
                  "fontFamily": "Gill Sans",
                  "fontSize": "14",
                  "letterSpacing": "normal",
                  "textAnchor": "middle",
                }
              }
              propList={
                [
                  "fill",
                ]
              }
              rotate={[]}
              x={
                [
                  "425.6189078043181",
                ]
              }
              y={[]}
            />
          </RNSVGText>
        </RNSVGGroup>
        <RNSVGGroup>
          <RNSVGText
            content={null}
            dx={
              [
                "0",
              ]
            }
            dy={[]}
            font={{}}
            name="chart-axis-1-tickLabels-3"
            rotate={[]}
            x={
              [
                "550.8252104057575",
              ]
            }
            y={
              [
                "472.97",
              ]
            }
          >
            <RNSVGTSpan
              content="400"
              dx={
                [
                  "0",
                ]
              }
              dy={
                [
                  "0",
                ]
              }
              fill={
                {
                  "payload": 4280624421,
                  "type": 0,
                }
              }
              font={
                {
                  "fontFamily": "Gill Sans",
                  "fontSize": "14",
                  "letterSpacing": "normal",
                  "textAnchor": "middle",
                }
              }
              propList={
                [
                  "fill",
                ]
              }
              rotate={[]}
              x={
                [
                  "550.8252104057575",
                ]
              }
              y={[]}
            />
          </RNSVGText>
        </RNSVGGroup>
        <RNSVGGroup>
          <RNSVGText
            content={null}
            dx={
              [
                "0",
              ]
            }
            dy={[]}
            font={{}}
            name="chart-axis-1-tickLabels-4"
            rotate={[]}
            x={
              [
                "676.0315130071968",
              ]
            }
            y={
              [
                "472.97",
              ]
            }
          >
            <RNSVGTSpan
              content="500"
              dx={
                [
                  "0",
                ]
              }
              dy={
                [
                  "0",
                ]
              }
              fill={
                {
                  "payload": 4280624421,
                  "type": 0,
                }
              }
              font={
                {
                  "fontFamily": "Gill Sans",
                  "fontSize": "14",
                  "letterSpacing": "normal",
                  "textAnchor": "middle",
                }
              }
              propList={
                [
                  "fill",
                ]
              }
              rotate={[]}
              x={
                [
                  "676.0315130071968",
                ]
              }
              y={[]}
            />
          </RNSVGText>
        </RNSVGGroup>
      </RNSVGGroup>
      <RNSVGGroup
        clipPath="victory-clip-2"
      >
        <RNSVGGroup>
          <RNSVGDefs>
            <RNSVGClipPath
              name="victory-clip-2"
            >
              <RNSVGRect
                height="400"
                vectorEffect={1}
                width="650"
                x="50"
                y="50"
              />
            </RNSVGClipPath>
          </RNSVGDefs>
        </RNSVGGroup>
      </RNSVGGroup>
      <View
        pointerEvents="box-none"
        style={
          {
            "left": 0,
            "position": "absolute",
            "top": 0,
            "zIndex": 99,
          }
        }
      >
        <RNSVGSvgView
          align="xMidYMid"
          bbHeight="500"
          bbWidth="750"
          focusable={false}
          height={500}
          meetOrSlice={0}
          minX={0}
          minY={0}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              {
                "height": "100%",
                "overflow": "visible",
                "width": "100%",
              },
              {
                "flex": 0,
                "height": 500,
                "width": 750,
              },
            ]
          }
          vbHeight={500}
          vbWidth={750}
          width={750}
        >
          <RNSVGGroup />
        </RNSVGSvgView>
      </View>
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;
